### PR TITLE
Add vertical CRS entries to srs db, and rework projection tree widget to use proper model

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR makes a number of changes to CRS handling across QGIS:

- Purely vertical crs are now added to the internal srs.db (previously they were skipped over). This enables them to be selected from the various CRS selection widgets.
- A one-time read of the srs db tbl_srs is now made, with the results being stored in memory for the session, instead of reading the table for every widget construction (and with every updated crs filter string). By avoiding the cost of re-reading the table we can greatly simplify QgsProjectionSelectionTreeWidget by dropping all the deferred loading code.
- A proper tree model is now used by QgsProjectionSelectionTreeWidget instead of a QTreeWidget, giving us a more performant + flexible tree view of crs.
- I've implemented a sort/filter proxy model for the crs tree model, so that we can easily filter by the available crs via a number of criteria

Ultimately, this opens the possibility for a CRS selector enabling users to pick from vertical crs, instead of just horizontal crs. Right now all CRS widgets will still default to showing just horizontal CRS (to match behavior of older QGIS releases), but this functionality will be used in follow up PRs to enable eg choice of vertical CRS for layers and projects.

### CODE REVIEW

**Comment by nyalldawson** [comment]
(Note: anyone who builds there own qgis will need to delete the /tmp/srs6.db file in order for this to be rebuilt with the vertical crs)

File: src/core/proj/qgscoordinatereferencesystem.cpp [hunk 1]
```diff
@@ -2642,24 +2724,27 @@ int QgsCoordinateReferenceSystem::syncDatabase()
 
       QString srsProj4;
       QString srsDesc;
+      QString dbSrsType;
```
Comment (nirvn) [inline src/core/proj/qgscoordinatereferencesystem.cpp:2727]: For the sake of consistency should we rename the other two variables to dbSrs* ?

```suggestion
      QString dbSrsProj4;
      QString dbSrsDesc;
      QString dbSrsType;
```

File: src/core/proj/qgscoordinatereferencesystem.cpp [hunk 2]
```diff
@@ -2714,11 +2799,12 @@ int QgsCoordinateReferenceSystem::syncDatabase()
...
                 .arg( QgsSqliteUtils::quotedString( code ) )
                 .arg( isGeographic ? 1 : 0 )
-                .arg( deprecated ? 1 : 0 );
+                .arg( deprecated ? 1 : 0 )
+                .arg( QgsSqliteUtils::quotedString( srsTypeString ) );
         }
         else
         {
-          sql = QStringLiteral( "INSERT INTO tbl_srs(description,projection_acronym,ellipsoid_acronym,parameters,srid,auth_name,auth_id,is_geo,deprecated) VALUES (%2,%3,%4,%5,%6,%7,%8,%9,%10)" )
+          sql = QStringLiteral( "INSERT INTO tbl_srs(description,projection_acronym,ellipsoid_acronym,parameters,srid,auth_name,auth_id,is_geo,deprecated,srs_type) VALUES (%2,%3,%4,%5,%6,%7,%8,%9,%10,%11)" )
```
Comment (nirvn) [inline src/core/proj/qgscoordinatereferencesystem.cpp:2807]: %1 is missing here (not a change you made but a preexisting condition), should we fix?

File: src/core/proj/qgscoordinatereferencesystemregistry.cpp [hunk 3]
```diff
@@ -431,3 +432,40 @@ QSet<QString> QgsCoordinateReferenceSystemRegistry::authorities() const
...
+  {
+    const QString srsDatabaseFileName = QgsApplication::srsDatabaseFilePath();
+    if ( QFileInfo::exists( srsDatabaseFileName ) )
+    {
+      // open the database containing the spatial reference data, and do a one-time read
+      sqlite3_database_unique_ptr database;
+      int result = database.open_v2( srsDatabaseFileName, SQLITE_OPEN_READONLY, nullptr );
+      if ( result == SQLITE_OK )
+      {
+        const QString sql = QStringLiteral( "select description, srs_id, auth_name, auth_id, projection_acronym, deprecated, srs_type from tbl_srs" );
```
Comment (nirvn) [inline src/core/proj/qgscoordinatereferencesystemregistry.cpp:449]:
```suggestion
        const QString sql = QStringLiteral( "SELECT description, srs_id, auth_name, auth_id, projection_acronym, deprecated, srs_type FROM tbl_srs" );
```

File: src/gui/qgscoordinatereferencesystemmodel.cpp [hunk 4]
```diff
@@ -0,0 +1,751 @@
...
+              return groupNode->name();
+
+            default:
+              break;
+          }
+          break;
+
+        case Qt::FontRole:
+        {
+          QFont font;
```
Comment (nirvn) [inline src/gui/qgscoordinatereferencesystemmodel.cpp:100]: Why not going through a QStyledItemDelegate for custom fonts? That way, if a font had been customized via CSS theme, you wouldn't be resetting the font's family/size. 

No strong feelings though.

File: src/gui/qgscoordinatereferencesystemmodel.cpp [hunk 5]
```diff
@@ -0,0 +1,751 @@
...
+{
+  QgsCoordinateReferenceSystemModelGroupNode *parentNode = mRootNode.get();
+  std::unique_ptr< QgsCoordinateReferenceSystemModelCrsNode > crsNode = std::make_unique< QgsCoordinateReferenceSystemModelCrsNode>( record );
+
+  QString groupName;
+  QString groupId;
+  QIcon groupIcon;
+  if ( record.authName == QLatin1String( "USER" ) )
+  {
+    groupName = tr( "User Defined Coordinate Systems" );
```
Comment (nirvn) [inline src/gui/qgscoordinatereferencesystemmodel.cpp:361]: User-defined or user defined? Also, should it be User-defined Coordinate >>Reference<< Systems?

File: src/gui/qgscoordinatereferencesystemmodel.cpp [hunk 6]
```diff
@@ -0,0 +1,751 @@
...
+  QgsCrsDbRecord userRecord;
+  userRecord.authName = QStringLiteral( "CUSTOM" );
+  userRecord.description = crs.description().isEmpty() ? tr( "Custom CRS" ) : crs.description();
+  userRecord.type = crs.type();
+
+  QgsCoordinateReferenceSystemModelGroupNode *group = mRootNode->getChildGroupNode( QStringLiteral( "CUSTOM" ) );
+  if ( !group )
+  {
+    std::unique_ptr< QgsCoordinateReferenceSystemModelGroupNode > newGroup = std::make_unique< QgsCoordinateReferenceSystemModelGroupNode >(
+          tr( "Custom Coordinate Systems" ),
```
Comment (nirvn) [inline src/gui/qgscoordinatereferencesystemmodel.cpp:471]: Is it a Custom Coordinate >>Reference<< System or a Custom Coordinate System? I'm asking as a few lines above, you're using "Custom C>>R<<S.

File: src/gui/qgscoordinatereferencesystemmodel.cpp [hunk 7]
```diff
@@ -0,0 +1,751 @@
...
+          QgsApplication::getThemeIcon( QStringLiteral( "/user.svg" ) ),  QStringLiteral( "CUSTOM" ) );
```
Comment (nirvn) [inline src/gui/qgscoordinatereferencesystemmodel.cpp:472]: Would be nice to work out nice icons for these groups (in a separate PR)

File: tests/src/python/test_qgscoordinatereferencesystemmodel.py [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+        registry = QgsApplication.coordinateReferenceSystemRegistry()
+        crs = QgsCoordinateReferenceSystem.fromProj("+proj=aea +lat_1=20 +lat_2=-23 +lat_0=4 +lon_0=29 +x_0=10 +y_0=3 +datum=WGS84 +units=m +no_defs")
+        res = registry.addUserCrs(crs, 'my custom crs')
+        self.assertEqual(res, 100000)
+
+        self.assertEqual(model.rowCount(QModelIndex()), prev_top_level_count + 1)
+        top_level_items = [
+            model.data(model.index(row, 0, QModelIndex()), Qt.DisplayRole) for row in range(model.rowCount(QModelIndex()))
+        ]
+        self.assertIn('User Defined Coordinate Systems', top_level_items)
```
Comment (nirvn) [inline tests/src/python/test_qgscoordinatereferencesystemmodel.py:229]: Note, if you change to user-defined, it'll need to be updated in this test too.

File: src/gui/qgscoordinatereferencesystemmodel.cpp (hunk 4 above)
Comment (nyalldawson) [reply src/gui/qgscoordinatereferencesystemmodel.cpp:100]: I think that's overkill personally -- the font role is exactly designed for this use case and is considerably simpler vs using a delegate

**Review by nirvn — APPROVED** [review]

**Pull request merged**